SDL/Quintuplet.cu

```diff
@@ -0,0 +1,646 @@
+    cudaMemset(modulesInGPU.quintupletModuleIndices, -1, sizeof(int) * (nLowerModules));
+
+    short* module_subdets;
+    cudaMallocHost(&module_subdets, nModules* sizeof(short));
+    cudaMemcpy(module_subdets,modulesInGPU.subdets,nModules*sizeof(short),cudaMemcpyDeviceToHost);
+    unsigned int* module_lowerModuleIndices;
+    cudaMallocHost(&module_lowerModuleIndices, nLowerModules * sizeof(unsigned int));
+    cudaMemcpy(module_lowerModuleIndices,modulesInGPU.lowerModuleIndices, nLowerModules * sizeof(unsigned int),cudaMemcpyDeviceToHost);
+    short* module_layers;
+    cudaMallocHost(&module_layers, nModules * sizeof(short));
+    cudaMemcpy(module_layers,modulesInGPU.layers,nModules * sizeof(short),cudaMemcpyDeviceToHost);
+    int* module_quintupletModuleIndices;
+    cudaMallocHost(&module_quintupletModuleIndices, nLowerModules * sizeof(int));
+    cudaMemcpy(module_quintupletModuleIndices,modulesInGPU.quintupletModuleIndices,nLowerModules *sizeof(int),cudaMemcpyDeviceToHost);
+    
+    unsigned int* nTriplets;
+    cudaMallocHost(&nTriplets, nLowerModules * sizeof(unsigned int));
+    cudaMemcpy(nTriplets, tripletsInGPU.nTriplets, nLowerModules * sizeof(unsigned int), cudaMemcpyDeviceToHost);
+
+    //start filling
+    for(unsigned int i = 0; i < nLowerModules; i++)
+    {
+        //condition for a quintuple to exist for a module
+        //TCs don't exist for layers 5 and 6 barrel, and layers 2,3,4,5 endcap
+        unsigned int idx = module_lowerModuleIndices[i];
+        if(((module_subdets[idx] == SDL::Barrel and module_layers[idx] < 5) or (module_subdets[idx] == SDL::Endcap and module_layers[idx] == 1)) and nTriplets[i] != 0)
+        {
+            module_quintupletModuleIndices[i] = nEligibleModules * maxQuintuplets;
+            indicesOfEligibleModules[nEligibleModules] = i;
+            nEligibleModules++;
+            maxTriplets = max(nTriplets[i], maxTriplets);
+            
+        }
+    }
+    cudaMemcpy(modulesInGPU.quintupletModuleIndices,module_quintupletModuleIndices,nLowerModules*sizeof(int),cudaMemcpyHostToDevice);
+    cudaMemcpy(modulesInGPU.nEligibleT5Modules,&nEligibleModules,sizeof(unsigned int),cudaMemcpyHostToDevice);
+    cudaFreeHost(module_subdets);
+    cudaFreeHost(module_lowerModuleIndices);
+    cudaFreeHost(module_layers);
+    cudaFreeHost(module_quintupletModuleIndices);
+    cudaFreeHost(nTriplets);
+}
+
+
+void SDL::createQuintupletsInUnifiedMemory(struct SDL::quintuplets& quintupletsInGPU, const unsigned int& maxQuintuplets, const unsigned int& nLowerModules, const unsigned int& nEligibleModules)
+{
+    unsigned int nMemoryLocations = maxQuintuplets * nEligibleModules;
+    std::cout<<"Number of eligible T5 modules = "<<nEligibleModules<<std::endl;
+
+#ifdef CACHE_ALLOC
+    cudaStream_t stream = 0;
+    quintupletsInGPU.tripletIndices = (unsigned int*)cms::cuda::allocate_managed(nMemoryLocations * 2 * sizeof(unsigned int));
```

Comment:
quintupletsInGPU.tripletIndices = (unsigned int*)cms::cuda::allocate_managed(nMemoryLocations * 2 * sizeof(unsigned int),stream);
You need the stream here